### PR TITLE
Test newline normalization in form-associated custom elements.

### PR DESCRIPTION
**Pull request by andreubotella**

### CODE REVIEW

**Comment by domenic** [comment]
@tkent-google will probably be the best person to review this, but from what I gather in IRC, you are working on a large project surrounding form data serialization and newlines, and so probably it'd be best to wait on reviews until that's settled and you have a writeup somewhere explaining the desired behavior?

**Comment by andreubotella** [comment]
Although I run into this as part of working on whatwg/url#562, these tests (other than the ones that deal with filenames) reflect the spec-mandated behavior: setting a string or a `File` as a FACE's value should cause the [entry construction algorithm](https://html.spec.whatwg.org/#face-entry-construction) to invoke [append an entry](https://html.spec.whatwg.org/#append-an-entry), which newline-normalizes names and string values, but setting a `FormData` should cause the entry construction algorithm to append the `FormData`'s entries without normalization.

Since the parts dealing with `File` objects will probably change, I could split them into a `.tentative` file.

**Comment by domenic** [comment]
In that case it's probably fine to review and merge as-is. That will make any future reviews for changes related to `File` objects simpler.

I will assign @tkent-google then, but @tkent-google feel free to say you would rather have someone else review; I realize you did not sign up to review this directory in general.

**Comment by andreubotella** [comment]
So the (bare) "filename" tests don't agree with the spec as it is right now, but would agree with any change to the spec that would fix whatwg/url#562. But the "FormData filename" tests might depend on the specifics of that fix. So on second thought, it'd be best to remove those or make them tentative for now.

File: custom-elements/form-associated/ElementInternals-setFormValue.html [hunk 1]
```diff
@@ -122,4 +122,296 @@
...
+      '<input name=name-pd1 value="value-pd1">' +
+      "<my-control name=a></my-control>" +
+      "</form>" +
+      '<iframe name="if1"></iframe>';
+  $("my-control").value = new File([], "b\n\rc");
+  return submitPromise(t).then((query) => {
+    assert_equals(query, "?name-pd1=value-pd1&a=b%0D%0A%0D%0Ac");
+  });
+}, "Newline normalization - \\n\\r in filename");
+
```
Comment (tkent-google) [inline custom-elements/form-associated/ElementInternals-setFormValue.html:273]: Repeating similar code is not great. Let's make a helper function like:

```javascript
function testSerializedEntry(name, expectedName, value, expectedValue, description) {
  promise_test(async t => {
    $("#container").innerHTML = '<form action="/common/blank.html" target="if1">' +
        '<my-control</my-control>' +
        '</form>' +
        '<iframe name="if1"></iframe>';
    $("my-control").setAttribute('name', name);
    $("my-control").value = value;
    let query = await submitPromise(t);
    assert_equals(query, `?${expectedName}=${expectedValue}`);
  }, `Newline normalization - ${description}');
}
testSerializedEntry('a', 'a', new File([], 'b\n\rc'), 'b%0D%0A%0D%0AC', '\\n\\r in filename');
```

File: custom-elements/form-associated/ElementInternals-setFormValue.html [hunk 2]
```diff
@@ -122,4 +122,296 @@
...
+  $("#container").innerHTML = '<form action="/common/blank.html" target="if1">' +
+      '<input name=name-pd1 value="value-pd1">' +
+      "<my-control></my-control>" +
+      "</form>" +
+      '<iframe name="if1"></iframe>';
+  $("my-control").setValues([["a", new File([], "b\n\rc")]]);
+  return submitPromise(t).then((query) => {
+    assert_equals(query, "?name-pd1=value-pd1&a=b%0A%0Dc");
+  });
+}, "Newline normalization - \\n\\r in FormData filename");
```
Comment (tkent-google) [inline custom-elements/form-associated/ElementInternals-setFormValue.html:416]: Ditto.

File: custom-elements/form-associated/ElementInternals-setFormValue.html [hunk 3]
```diff
@@ -122,4 +122,296 @@
     assert_equals(query, '?name-pd1=value-pd1');
   });
 }, 'setFormValue with an empty FormData should submit nothing');
+
+promise_test((t) => {
+  $("#container").innerHTML = '<form action="/common/blank.html" target="if1">' +
```
[the remark this answers is not in the document]
Comment (andreubotella) [reply custom-elements/form-associated/ElementInternals-setFormValue.html]: That was my IDE's autoformatting. I noticed and undid the indentation changes but I missed the quotes.

**Review by tkent-google — APPROVED** [review]

**Pull request merged**